### PR TITLE
v7 Simplify Web Test Runner config and CLI

### PR DESCRIPTION
**Pull request by Sakchai-Refinitiv**
## Description

v7 Simplify Web Test Runner config and CLI

- Prevent yargs CLI options crash with Web Test Runner CLI by using Node API
- Simplify CLI options by using sharable options for package and element CLI
- Simplify Web Test Runner config by moving yargs CLI out of the configs
- Show element name on BrowserStack if running test on a specific element

## Type of change

- [x] Refactor (improves code without changing its functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: scripts/tests/run.mjs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+const testAll = packageName === 'all' || packageName === undefined;
+const basePath = path.join(PACKAGES_ROOT, testAll ? '*' : packageName);
+const watch = argv.watch;
+const snapshots = argv.updateSnapshots;
+const browserstack = argv.browserstack && !watch;
+const testCoverage = argv.includeCoverage;
+let testTarget = packageName;
+
+// Environment variables to use for overriding config script
+env.testCoverage = testCoverage;
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]: I don't think we need to use environment variable here. If the export has to be dynamic, we could export a function with a param instead.

File: scripts/tests/run.mjs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+// Test on BrowserStack`
+if (browserstack) {
+  const sharedCapabilities = {
+    'browserstack.user': env.BROWSERSTACK_USERNAME,
+    'browserstack.key': env.BROWSERSTACK_ACCESS_KEY,
+    project: env.BROWSERSTACK_PROJECT_NAME || 'Refinitiv UI',
+    name: testTarget,
+    build: `build ${env.BROWSERSTACK_BUILD || 'unknown'}`,
+    timeout: 1800, // Maximum
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]: intead of "maximum" implied by the prop name `timeout`, we could provide the number unit e.g. time in milliseconds

File: scripts/tests/run.mjs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  const launchers = [];
+  argv.browserstack.forEach((option) => {
+    switch (option) {
+      case 'default':
+        BrowserStack.defaultBrowsers.forEach(browser => launchers.push(BrowserStack.config[browser]));
+        break;
+      case 'supported':
+        BrowserStack.supportedBrowsers.forEach(browser => {
+          // Disable testing on Safari, we have to check all test cases are passed before enabling it again
+          if (!browser.includes('safari')) {
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]: we should check for `safari_minus1` & `safari_minus2` as well. Also, we're going to enable testing on Safari this sprint. Does it mean this check would be removed very soon?

File: scripts/tests/run.mjs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    build: `build ${env.BROWSERSTACK_BUILD || 'unknown'}`,
+    timeout: 1800, // Maximum
+  };
+
+  // Add BrowserStack launchers to config
+  const launchers = [];
+  argv.browserstack.forEach((option) => {
+    switch (option) {
+      case 'default':
+        BrowserStack.defaultBrowsers.forEach(browser => launchers.push(BrowserStack.config[browser]));
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs:81]: why don't we exclude Safari in case of `default`? Plus, safari could be passed as `browserstack` argument direcly. We haven't handled such case.

File: scripts/tests/run.mjs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  });
+
+  // Add BrowserStack launchers to config
+  const browsers = [];
+  launchers.forEach(launcher => {
+    browsers.push(browserstackLauncher({ capabilities: { ...sharedCapabilities, ...launcher } }));
+  })
+
+  config.browsers = browsers; // Set all browsers to use BrowserStack
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs:98]: "Add BrowserStack launchers to config" should apply to this whole section (line 96 to line 101). "Set all browsers to use BrowserStack" seems unrelated and confusing here.

File: scripts/tests/run.mjs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  testTarget = target;
+  env.testElement = target;
+} else if (packageName === 'elements') {
+  env.testElement = 'all';
+}
+
+const config = {
+  ...wtrConfig,
+  files: [path.join(basePath , '/__test__/**/*.test.js')],
+  watch: argv.watch,
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]:
```suggestion
  watch
```

File: scripts/tests/run.mjs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  const browsers = [];
+  launchers.forEach(launcher => {
+    browsers.push(browserstackLauncher({ capabilities: { ...sharedCapabilities, ...launcher } }));
+  })
+
+  config.browsers = browsers; // Set all browsers to use BrowserStack
+}
+
+// Specify browser to run the unit test & convert browser naming to playwright's one
+const optionBrowser = argv.browsers;
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]: let's move this to line 25 section where arguments are extracted into variables.

File: scripts/tests/run.mjs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// Strip argv (options) out to prevent them pass to the web-test-runner cli
+process.argv = process.argv.slice(0, 2);
+
+info(watch ? `Start Dev Server: ${ testTarget }` : `Test: ${ testTarget }`);
+
+if (snapshots) {
+  info(`Update and prune snapshots: ${ testTarget }`);
+  process.argv.push('--update-snapshots');
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs:128]: Could you add a comment explaining that `update-snapshots` must be passed as  `process.argv` only?

File: scripts/tests/run.mjs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        break;
+      default:
+        browser = configBrowser.product;
+        break;
+    }
+    return optionBrowser.includes(browser);
+  });
+}
+
+// Strip argv (options) out to prevent them pass to the web-test-runner cli
```
Comment (wattachai-lseg) [inline scripts/tests/run.mjs]:
```suggestion
// Strip argv (options) out to prevent web-test-runner picking them up
```

File: scripts/tests/cli-options.mjs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env node
+
+import { BrowserStack, DefaultBrowsers } from '../../browsers.config.mjs';
+
+/**
+ * Use standard options for the yargs CLI
+ * @param {Argv} yargs argv
+ * @returns {Argv} yargs argv
+ */
+export const useTestOptions = (argv) => {
```
Comment (wattachai-lseg) [inline scripts/tests/cli-options.mjs]: this looks like a builder function. the first param is `yargs` not `argv`

File: packages/elements/scripts/cmd/test.mjs [hunk 11]
```diff
@@ -1,94 +1,36 @@
...
-      type: 'string',
-      alias: 'o',
-      default: 'full',
-      choices: ['full', 'minimal'],
-      description: 'Print output to the console'
-    })
     .completion('completion', () => elements);
+
+  // Use shared test options for the CLI
+  useTestOptions(yargs);
```
Comment (wattachai-lseg) [inline packages/elements/scripts/cmd/test.mjs:21]: missing `return` statement
```suggestion
  return useTestOptions(yargs);
```

File: packages/elements/scripts/cmd/test.mjs [hunk 12]
```diff
@@ -1,94 +1,36 @@
...
-  const browsers = argv.browsers ? argv.browsers.join(' '): '';
-  const browserstack = argv.browserstack ? argv.browserstack.join(' ') : '';
-
-  info(watch ? `Start Dev Server: ${ element }` : `Test: ${ element }`);
-
-  if (snapshots) {
-    info(`Update and prune snapshots: ${ element }`);
-  }
+  // Remove command and forward params only to main test file
+  let params = hideBin(process.argv).slice(1)
```
Comment (wattachai-lseg) [inline packages/elements/scripts/cmd/test.mjs:25]: yargs provides arguments for us with `argv` param. We should use it instead.

File: cli.mjs [hunk 13]
```diff
@@ -8,6 +8,7 @@ import { PACKAGES_ROOT, errorHandler, getJSON } from './scripts/helpers/esm.mjs'
 
 const argvNoBin = hideBin(process.argv);
 const argv = yargs(argvNoBin)
+  .scriptName("refinitiv-ui")
```
Comment (wattachai-lseg) [inline cli.mjs]: I don't think there is any command with such name.

File: packages/elements/web-test-runner.config.mjs
Comment (wattachai-lseg) [inline packages/elements/web-test-runner.config.mjs:1]: it seems this file is not in use. Did I miss something?

**Review by wattachai-lseg — CHANGES_REQUESTED** [review]
I have quite a number of concerns. Hope we can resolve them swiftly.

File: packages/elements/web-test-runner.config.mjs
Comment (Sakchai-Refinitiv) [reply packages/elements/web-test-runner.config.mjs:1]: This config is use for running test element, it inherits options from the parent config

Comment (wattachai-lseg) [reply packages/elements/web-test-runner.config.mjs:1]: Could you point me to its usage location?

File: packages/elements/scripts/cmd/test.mjs (hunk 11 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/scripts/cmd/test.mjs:21]: This method just modifies the object, so it does not need to return.

File: packages/elements/scripts/cmd/test.mjs (hunk 12 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/scripts/cmd/test.mjs:25]: The argv form yargs is already converted to an object that makes it harder to forward to the child process

File: scripts/tests/cli-options.mjs (hunk 10 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/cli-options.mjs]: If you see the type in yargs it is `Yargs.Argv` interface

File: scripts/tests/run.mjs (hunk 1 above)
Comment (wattachai-lseg) [reply scripts/tests/run.mjs]: need to use it as WTR would search for its config in `elements` package. Environment variable would be the only way to pass value to the config.
Let's mention `elements` package here in the comment.

File: cli.mjs (hunk 13 above)
Comment (Sakchai-Refinitiv) [reply cli.mjs]: This is optional I add it to prevent confusion because we have the two cli.mjs files

File: packages/elements/web-test-runner.config.mjs
Comment (Sakchai-Refinitiv) [reply packages/elements/web-test-runner.config.mjs:1]: Behavior is the same test group https://modern-web.dev/docs/test-runner/cli-and-configuration/#test-groups

File: scripts/tests/run.mjs (hunk 4 above)
Comment (wattachai-lseg) [reply scripts/tests/run.mjs:81]: Testing and ensuring green test result of Safari would be tracked separately. Marking as resolved.

File: scripts/tests/run.mjs (hunk 1 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Updated a comment.

File: scripts/tests/run.mjs (hunk 3 above)
Comment (wattachai-lseg) [reply scripts/tests/run.mjs]: Testing and ensuring green test result of Safari would be tracked separately. Marking as resolved.

File: scripts/tests/run.mjs (hunk 2 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Removed

File: scripts/tests/run.mjs (hunk 3 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Removed

File: scripts/tests/run.mjs (hunk 4 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs:81]: Updated, Safari included to the default browsers in v7.

File: scripts/tests/run.mjs (hunk 6 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Fixed

File: scripts/tests/run.mjs (hunk 7 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Moved

File: scripts/tests/run.mjs (hunk 8 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs:128]: Web Test Runner does not provide a config to update snapshots, so the CLI option is the only way.  I already added a code comment.

File: scripts/tests/run.mjs (hunk 9 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: Updated

File: scripts/tests/cli-options.mjs (hunk 10 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/cli-options.mjs]: Renamed to yargs

File: scripts/tests/run.mjs (hunk 5 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs:98]: Update comment.

File: scripts/tests/run.mjs (hunk 2 above)
Comment (wattachai-lseg) [reply scripts/tests/run.mjs]: this `timeout` is optional?

File: cli.mjs (hunk 13 above)
Comment (wattachai-lseg) [reply cli.mjs]: image

the upper one show a result with `scriptName`. The lower one show a result without `scriptName`.

We should remove this `scriptName()` as it only creates confusion to anyone using `--help` option. 
While `node refinitiv-ui test core` will never work, `node cli.mjs test core` will do the job.

**Review by wattachai-lseg — CHANGES_REQUESTED** [review]
only 2 more locations.

File: scripts/tests/run.mjs (hunk 2 above)
Comment (Sakchai-Refinitiv) [reply scripts/tests/run.mjs]: The `BrowserstackLauncher` class always overwrites `timeout` to 300 which means it does not need `timeout` here.

File: cli.mjs (hunk 13 above)
Comment (Sakchai-Refinitiv) [reply cli.mjs]: That's right. removed

**Review by wattachai-lseg — APPROVED** [review]
LGTM
